From 9be57dd07a489e681b58328385aac5bd1ec5945e Mon Sep 17 00:00:00 2001
From: Joey Riches <josephriches@gmail.com>
Date: Fri, 18 Sep 2026 22:09:01 +0100
Subject: [PATCH 1/1] daemon: Add support for default user groups

Similarly to extra_admin_groups, add support for default_user_groups.
When set, whenever a user is created (regardless of user type) they will
be added to these groups automatically.

This is provided to allow distributions to ensure users have a
baseline level of desired functionality. Typically, only users created
at installation time would be added to all desired groups. However,
users created post installation would be sod outta luck which would
require users to go into the terminal and add themselves to the desired
groups.

An alternative would be to patch shadow
(see https://github.com/shadow-maint/shadow/issues/160) but admins who
are modifying users from the terminal are most likely competent enough
to understand what groups they wish users to be part of.

---
 meson.build       |  4 ++++
 meson_options.txt |  1 +
 src/daemon.c      | 46 ++++++++++++++++++++++++++++++----------------
 3 files changed, 35 insertions(+), 16 deletions(-)

diff --git a/meson.build b/meson.build
index 1f06aa3..a10bb92 100644
--- a/meson.build
+++ b/meson.build
@@ -220,9 +220,12 @@ if admin_group == ''
   endif
 endif
 
+default_user_groups = ','.join(get_option('default_user_groups'))
+
 extra_admin_groups = ','.join(get_option('extra_admin_groups'))
 
 config_h.set_quoted('ADMIN_GROUP', admin_group)
+config_h.set_quoted('DEFAULT_USER_GROUPS', default_user_groups)
 config_h.set_quoted('EXTRA_ADMIN_GROUPS', extra_admin_groups)
 
 config_h.set('MINIMUM_UID', get_option('minimum_uid'))
@@ -274,6 +277,7 @@ meson.add_install_script(
 output = '\n' + meson.project_name() + ' was configured with the following options:\n'
 output += '** DocBook documentation build: ' + enable_docbook.to_string() + '\n'
 output += '** Administrator group: ' + admin_group + '\n'
+output += '** Default user groups: ' + default_user_groups + '\n'
 output += '** Extra administrator groups: ' + extra_admin_groups + '\n'
 output += '** GDM configuration: ' + gdm_conf_file + '\n'
 output += '** LightDM configuration: ' + lightdm_conf_file + '\n'
diff --git a/meson_options.txt b/meson_options.txt
index 5524ea0..4afbf78 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -5,6 +5,7 @@ option('wtmpfile', type: 'string', value: '', description: 'override filepath of
 
 option('admin_group', type: 'string', value: '', description: 'Set group for administrative accounts')
 option('extra_admin_groups', type: 'array', value: [], description: 'Comma-separated list of extra groups that administrator users are part of')
+option('default_user_groups', type: 'array', value: [], description: 'Comma-separated list of groups that all users are part of by default')
 option('minimum_uid', type: 'integer', value: 1000, description: 'Set minimum uid for human users')
 
 option('elogind', type: 'boolean', value: false, description: 'Use elogind')
diff --git a/src/daemon.c b/src/daemon.c
index 921fa05..e542333 100644
--- a/src/daemon.c
+++ b/src/daemon.c
@@ -1753,7 +1753,7 @@ create_data_free (gpointer data)
 static User *
 daemon_create_user_backend (Daemon         *daemon,
                             CreateUserData *cd,
-                            GStrv           admin_groups,
+                            GStrv           groups,
                             GError        **error)
 {
         const char *password = ""; /* homed doesn't support passwordless users yet, so we set an empty passwd instead */
@@ -1778,10 +1778,10 @@ daemon_create_user_backend (Daemon         *daemon,
         json_object_object_add (obj, "lastChangeUSec", json_object_new_uint64 (now));
         json_object_object_add (obj, "lastPasswordChangeUSec", json_object_new_uint64 (now));
 
-        if (admin_groups != NULL) {
+        if (groups != NULL) {
                 g_autoptr (json_object) member_of = json_object_new_array ();
-                for (gsize i = 0; admin_groups[i] != NULL; i++) {
-                        json_object_array_add (member_of, json_object_new_string (admin_groups[i]));
+                for (gsize i = 0; groups[i] != NULL; i++) {
+                        json_object_array_add (member_of, json_object_new_string (groups[i]));
                 }
                 json_object_object_add (obj, "memberOf", g_steal_pointer (&member_of));
         }
@@ -1826,10 +1826,10 @@ daemon_create_user_backend (Daemon         *daemon,
 static User *
 daemon_create_user_backend (Daemon         *daemon,
                             CreateUserData *cd,
-                            GStrv           admin_groups_array,
+                            GStrv           groups_array,
                             GError        **error)
 {
-        g_autofree char *admin_groups = NULL;
+        g_autofree char *groups = NULL;
         GError *inner_error = NULL;
         User *user = NULL;
         const gchar *argv[9] = { NULL };
@@ -1840,11 +1840,11 @@ daemon_create_user_backend (Daemon         *daemon,
         argv[i++] = "-c";
         argv[i++] = cd->real_name;
 
-        if (admin_groups_array != NULL) {
-                admin_groups = g_strjoinv (",", admin_groups_array);
+        if (groups_array != NULL) {
+                groups = g_strjoinv (",", groups_array);
 
                 argv[i++] = "-G";
-                argv[i++] = admin_groups;
+                argv[i++] = groups;
         }
 
         argv[i++] = "--";
@@ -1871,7 +1871,8 @@ daemon_create_user_authorized_cb (Daemon                *daemon,
                                   gpointer               data)
 {
         g_autoptr (GError) error = NULL;
-        g_auto (GStrv) admin_groups = NULL;
+        g_auto (GStrv) groups = NULL;
+        g_autoptr (GStrvBuilder) groups_builder = g_strv_builder_new ();
 
         CreateUserData *cd = data;
         User *user = NULL;
@@ -1881,21 +1882,30 @@ daemon_create_user_authorized_cb (Daemon                *daemon,
                 return;
         }
 
-        if (cd->account_type == ACCOUNT_TYPE_ADMINISTRATOR) {
-                g_autoptr (GStrvBuilder) admin_groups_builder = g_strv_builder_new ();
+        /* Add the default user groups that exist */
+        if (DEFAULT_USER_GROUPS != NULL && DEFAULT_USER_GROUPS[0] != '\0') {
+                g_auto (GStrv) default_user_groups = NULL;
+                default_user_groups = g_strsplit (DEFAULT_USER_GROUPS, ",", 0);
+                for (gsize i = 0; default_user_groups[i] != NULL; i++) {
+                        if (getgrnam (default_user_groups[i]) != NULL)
+                                g_strv_builder_add (groups_builder, default_user_groups[i]);
+                        else
+                                g_warning ("Default user group %s doesn’t exist: not adding the user to it", default_user_groups[i]);
+                }
+        }
 
-                g_strv_builder_add (admin_groups_builder, ADMIN_GROUP);
+        if (cd->account_type == ACCOUNT_TYPE_ADMINISTRATOR) {
+                g_strv_builder_add (groups_builder, ADMIN_GROUP);
                 if (EXTRA_ADMIN_GROUPS != NULL && EXTRA_ADMIN_GROUPS[0] != '\0') {
                         g_auto (GStrv) extra_admin_groups = NULL;
                         extra_admin_groups = g_strsplit (EXTRA_ADMIN_GROUPS, ",", 0);
                         for (gsize i = 0; extra_admin_groups[i] != NULL; i++) {
                                 if (getgrnam (extra_admin_groups[i]) != NULL)
-                                        g_strv_builder_add (admin_groups_builder, extra_admin_groups[i]);
+                                        g_strv_builder_add (groups_builder, extra_admin_groups[i]);
                                 else
                                         g_warning ("Extra admin group %s doesn’t exist: not adding the user to it", extra_admin_groups[i]);
                         }
                 }
-                admin_groups = g_strv_builder_end (admin_groups_builder);
                 sys_log (context, "create admin user '%s'", cd->user_name);
         } else if (cd->account_type == ACCOUNT_TYPE_STANDARD) {
                 sys_log (context, "create standard user '%s'", cd->user_name);
@@ -1904,7 +1914,11 @@ daemon_create_user_authorized_cb (Daemon                *daemon,
                 return;
         }
 
-        user = daemon_create_user_backend (daemon, cd, admin_groups, &error);
+        groups = g_strv_builder_end (groups_builder);
+        if (groups[0] == NULL)
+                g_clear_pointer (&groups, g_strfreev);
+
+        user = daemon_create_user_backend (daemon, cd, groups, &error);
 
         if (error == NULL)
                 accounts_accounts_complete_create_user (NULL, context, user_get_object_path (user));
-- 
2.55.0

